Quilt Registry Builder

Leo40Git opened the pull request:
Add registry builders!
- The base `QuiltRegistryBuilder` class
- Built-in (code-driven) registries: `QuiltBuiltinRegistryBuilder`
- Dynamic (data-driven) registries: `QuiltDynamicRegistryBuilder`
  - Registers into the `DynamicRegistryManager`!

### TODO
- [ ] Document dynamic registries more!!!
- [ ] Add dynamic registry testmod

Leo40Git committed:
Add QuiltRegistryBuilder

Leo40Git committed:
Add testmod for builder

QuiltRegistryBuilder: remove API note, since we DO get validated by Minecraft itself

Leo40Git committed:
QuiltRegistryBuilder: support dynamic registries

Leo40Git committed:
QuiltRegistryBuilder: clarify builtin/dynamic method side-effects

EnnuiL reviewed (approved):
I like this!

TheGlitch76 committed:
Merge branch '1.19' into feature/registry-builder

Leo40Git committed:
QuiltRegistryBuilder: unmangle builtin() javadoc

Leo40Git committed:
Merge branch '1.19' into feature/registry-builder

LambdAurora commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java:
```diff
@@ -0,0 +1,311 @@
+	 * @throws IllegalArgumentException if this builder is configured with a {@linkplain #syncBehavior(SyncBehavior) synchronization behavior} that is not {@link SyncBehavior#SKIPPED}
+	 */
+	@Contract("-> this")
+	public @NotNull QuiltRegistryBuilder<T> dynamic() {
```
So, I looked a bit into dynamic registries and to make this fully working we would need:
 - register the registry into [`BuiltinRegistries.REGISTRIES`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/BuiltinRegistries.html#REGISTRIES)
 - ask for a codec and a network codec (which may be the same), and put those into [`DynamicRegistryManager.INFOS`](https://maven.quiltmc.org/repository/release/org/quiltmc/quilt-mappings/1.19.2+build.12/quilt-mappings-1.19.2+build.12-javadoc.jar/net/minecraft/util/registry/DynamicRegistryManager.html#INFOS)
 
 And I believe that's it.

Leo40Git replied:
I think I'd prefer splitting the class into a `QuiltBuiltinRegistryBuilder` and a `QuiltDynamicRegistryBuilder`.

LambdAurora replied:
Sure, and have a base builder class to extend, sounds fine.

LambdAurora reviewed (commented):
Would be neat to also have some kind of `withBootstrap` method which takes a `Consumer<Registry>` so you can register some default values as soon as the registry is built, similar to some of Vanilla registries.

Leo40Git committed:
Groundwork for dynamic registry builder

Rename builder setters to "withXyz" format

Leo40Git committed:
QuiltRegistryBuilder: add withBootstrap

Leo40Git committed:
QuiltBuiltinRegistryBuilder: remove useless javadoc

Leo40Git committed:
Add QuiltDynamicRegistryBuilder

Leo40Git committed:
Fix up DynamicRegistryManager mixins

Leo40Git commented:
Marking as WIP/draft until 1) dynamic registry testmod is done and 2) dynamic registry JSON path is cleared up.

Leo40Git commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltDynamicRegistryBuilder.java:
```diff
@@ -0,0 +1,66 @@
+
+		BuiltinRegistries.register((Registry<Registry<Object>>) BuiltinRegistries.REGISTRIES, this.key.getValue(), (Registry<Object>) registry);
+
+		var infos = DynamicRegistryManager.INFOS;
+
+		infos.put(this.key, new DynamicRegistryManager.Info<>(this.key, this.entryCodec, this.networkEntryCodec));
```
```suggestion
		DynamicRegistryManager.INFOS.put(this.key, new DynamicRegistryManager.Info<>(this.key, this.entryCodec, this.networkEntryCodec));
```

Leo40Git committed:
Remove redundant var

Leo40Git committed:
Merge branch '1.19' into feature/registry-builder

Leo40Git committed:
Merge branch '1.19' into feature/registry-builder

Leo40Git commented:
Undrafting, but marking as `s: not working` due to the missing testmod (which will be worked on soon-ish!)

Dynamic registries need more documentation, as well.

Leo40Git committed:
check! style!

Leo40Git committed:
Return Registry instead of SimpleRegistry

Leo40Git committed:
Take RegistryKey instead of Identifier

This encourages creating a RegistryKey field for your registry,
and also neatly bypasses the requirement to specify the registry's entry type
using Java's weird "specify method type parameter" syntax.

Moved QuiltBuiltinRegistryBuilder test to its own package
Moved RegistryEntry from QuiltBuiltinRegistryBuilder test to its own file (now called GasType)

Leo40Git committed:
Move QuiltBuiltinRegistryBuilder.SyncBehavior enum to its own file

Leo40Git committed:
QuiltRegistryBuilder.withCustomHolderProvider: update example snippet

Leo40Git committed:
Merge branch '1.19' into feature/registry-builder

Leo40Git committed:
Merge branch '1.19' into feature/registry-builder

Leo40Git commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java:
```diff
@@ -0,0 +1,264 @@
+	 *     public static final RegistryKey<Registry<Thing>> THING_KEY = RegistryKey.ofRegistry(YourCoolMod.id("thing"));
+	 *
+	 *     public static final SimpleRegistry<Thing> THING =
+	 *         new QuiltRegistryBuilder(THING_KEY)
```
```suggestion
	 *         QuiltRegistryBuilder.builtin(THING_KEY)
```

Leo40Git committed:
Fix withCustomHolderProvider example snippet

Leo40Git commented on library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/QuiltRegistryBuilder.java:
```diff
@@ -0,0 +1,264 @@
+	 * This should be used if your registry entry needs to store its own holder. Add the following to your registry entry class:
+	 * <pre><code>
+	 * public class Thing {
+	 *     private final Holder<Thing> registryHolder = YourCoolModRegistries.THINGS.createIntrusiveHolder(this);
```
```suggestion
	 *     private final Holder<Thing> registryHolder = YourCoolModRegistries.THING.createIntrusiveHolder(this);
```

Leo40Git committed:
Fix withCustomHolderProvider example snippet, pt. 2

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/registry-builder

Leo40Git committed:
dynamicUnsynced -> dynamicUnsynchronized

EnnuiL commented:
Partially superseded by PR #271, but not to a point where this PR becomes completely useless! In fact, it would be a great complement to that PR, and yeah, it no longer has to worry about the whole dynamic registry dilemma
